### PR TITLE
Fix for cleanwallettransactions cleaning out immature coins

### PR DESCRIPTION
**Pull request by webworker01**
The method used for listunspent excludes immature coins. `cleanwallettransactions` does not need all the data and checks from `AvailableCoins` (aka listunspent) method as it just needs the lowest blockheight below which it can clear all transactions from the wallet. 

Updated to use a simplified loop (same as what's used later in the method) through wallet transactions and basic checks against the transactions/utxos - `CheckFinalTx` and orphaned

Also assert pwalletMain exists earlier as it was used for unlocking transactions during cleanup.

Diff without whitespace cleanup https://github.com/KomodoPlatform/komodo/pull/553/files?w=1

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -1155,19 +1157,36 @@ UniValue cleanwallettransactions(const UniValue& params, bool fHelp, const CPubK
         // get all locked utxos to relock them later.
         vector<COutPoint> vLockedUTXO;
         pwalletMain->ListLockedCoins(vLockedUTXO);
-        // unlock all coins so that the following call containes all utxos.
+        // unlock all coins so that the following call contains all utxos.
```
Comment (jmjatlanta) [inline src/wallet/rpcwallet.cpp:1160]: With using `pwalletMain->mapWallet`, unlocking and then re-locking is no longer necessary.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -1155,19 +1157,36 @@ UniValue cleanwallettransactions(const UniValue& params, bool fHelp, const CPubK
...
-              oldestTxDepth = out.nDepth;
+            const uint256& wtxid = it->first;
+            const CWalletTx* pcoin = &(*it).second;
+
+            //Check if transaction will be final in the next block to be created.
+            if (!CheckFinalTx(*pcoin))
+                continue;
+
+            //orphaned coins get cleaned
+            int nDepth = pcoin->GetDepthInMainChain();
```
Comment (jmjatlanta) [inline src/wallet/rpcwallet.cpp:1175]: This is a somewhat heavy call. A possible optimization is to temporarily store this data to make it unnecessary to call this again in the removal loop below. (scope creep, I know.)

**Review by jmjatlanta — CHANGES_REQUESTED** [review]
Thank you for your contribution. My request is simply to remove the unnecessary assert. The others are simply optimizations that you may consider.

A longer-winded restatement of the problem this PR fixes:

The RPC method `cleanwallettransactions` was too aggressive, in that it did not consider immature transactions. 

Previously it used the wallet's `AvailableCoins` method, which filters out immature transactions. When `cleanwallettransactions` calculated depth, it could calculate a depth that is too shallow, thereby removing transactions from the wallet that it should not have.

The new way uses the guts of the `AvailableCoins` method, but considers immature transactions when calculating the depth.

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (webworker01) [reply src/wallet/rpcwallet.cpp:1175]: I see, I read `GetDepthInMainChainINTERNAL` and think I understand what you're saying. I would like to attempt to solve this issue but will need to work on it next week.

Comment (jmjatlanta) [reply src/wallet/rpcwallet.cpp:1175]: If you do decide to tackle this, you may wish to do it in a separate PR. It is a performance fix, not a bug fix, and truly scope-creep. I have the bad habit of trying to fix everything when I should only fix 1 thing.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1119,10 +1119,12 @@ UniValue cleanwallettransactions(const UniValue& params, bool fHelp, const CPubK
             + HelpExampleRpc("cleanwallettransactions","\"1075db55d416d3ca199f55b6084e2115b9345e16c5cf302fc80e9d5fbf5d48d\"")
         );
 
+    assert(pwalletMain != NULL);
```
[the remark this answers is not in the document]
Comment (webworker01) [reply src/wallet/rpcwallet.cpp]: Fixed in c1576dc

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (webworker01) [reply src/wallet/rpcwallet.cpp]: Removed locking/unlocking in a6ac2e7

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (webworker01) [reply src/wallet/rpcwallet.cpp:1168]: It's all good! I believe I handled the other two things. I believe the original author had added in the unlocking to get a full list of transactions which as you said seems like it's not required. 

Builds on my system after these 2 changes, will try it out on one of my servers.

**Comment by dimxy** [comment]
I have a question about skipping txns with CheckFinalTx(*pcoin).
Could we have extra cleaned txns because of that?
Suppose there are several spent utxos and only one unspent utxo U1 in my wallet.
Let's I spend that uxto U1 with a tx TX2 which is not final yet.
if I call cleanwallettransactions after that this call would skip TX2' utxos but would see U1 as spent.
So the oldestTxDepth would still be 0 in this case and this may cause all txns in the wallet to get cleaned.

**Comment by webworker01** [comment]
I extracted this code from `AvailableCoins()` https://github.com/KomodoPlatform/komodo/blob/master/src/wallet/wallet.cpp#L3299 with the goal of not introducing too many changes at once. I don't think it will clean any more than before, ideally it will clean out less.

If my basic understanding of non-final transactions is correct, the tx in question would not be mined and have no depth to be considered. Please let me know if I understand this correctly.

**Comment by dimxy** [comment]
I am not sure this had been implemented correctly beforehand.
Basically a tx may be non-final for a certain moment if it has the nLockTime property bigger than the current chain timestamp (or height) plus it has a vin nSequence set to non-default value. In the KMD chain we create all txns final by default so this may not be an issue there (but in Tokel they can use txns that may be non-final to unlock time-locked utxos).
I ll try to create a sample to verify what I wrote.

**Comment by dimxy** [comment]
I tried to test my assumption that non final transaction in the wallet may affect tx cleaning. But I was wrong and the code apparently works okay with non-final txns.
I created a tx with nLockTime set in future and it was added okay to the wallet. However in the loop where the oldestTxDepth is calculated, the output of the previous tx which was referred by the non-final tx was determined as 'not-spent' (I feared that it would be treated as 'spent' and that previous tx would be removed)

**Review by dimxy — APPROVED** [review]

**Comment by ca333** [comment]
in archive for further eval - getting back ref. dpow2.0
